interfaces/builtin/udev: add spec to disable udev + device cgroup

anonymouse64 opened the pull request:
This enables an interface to declare that the snap plugging this interface controls it's own device cgroup. This is the case for the greengrass snap and may be the case for other container management snaps developed in the future. Ideally these snaps would be able to "stack" their cgroups inside the device cgroup enforced by snapd when a relevant interface is plugged, but that is not the case currently. 

I added some testing that is specific to the greengrass-support interface but I'm not sure what else should be tested for this feature so please let me know if I should add any more tests for this. 

~This is based on top of https://github.com/snapcore/snapd/pull/6162~ - this doesn't include the changes from #6162 anymore

zyga commented on interfaces/builtin/common.go:
```diff
@@ -162,8 +163,14 @@ func (iface *commonInterface) SecCompConnectedPlug(spec *seccomp.Specification,
-	for _, rule := range iface.connectedPlugUDev {
-		spec.TagDevice(rule)
+	// don't tag devices if the interface controls it's own device cgroup
+	if !iface.controlsDeviceCgroup {
```
Nitpick: could you please drop the negation and simply swap the if / else code inside.

anonymouse64 replied:
Done

zyga commented on interfaces/udev/spec.go:
```diff
@@ -44,6 +44,14 @@ type Specification struct {
+
+// ControlsDeviceCgroup marks a specification as needing to control
+// it's own device cgroup which prevents generation of any udev tagging rules
+// for this security tag
```
s/security tag/snap name/

anonymouse64 replied:
Done

zyga commented on interfaces/udev/spec.go:
```diff
@@ -93,6 +103,14 @@ func (c byTagAndSnippet) Less(i, j int) bool {
 
 // Snippets returns a copy of all the snippets added so far.
 func (spec *Specification) Snippets() (result []string) {
+	// if the interface controls it's own device cgroup, then
```
```suggestion
	// If the interface controls it's own device cgroup, then
```

anonymouse64 commented:
I rebased this PR back on master so as not to include the greengrass-support changes in #6162 as requested, but note that either this PR will need to be rebased after that PR is merged or vice versa as the two patches don't apply cleanly to each other

anonymouse64 commented:
Oh also we discussed this on the forum and in a consequent hangout meeting here: https://forum.snapcraft.io/t/allowing-interfaces-to-disable-device-cgroup-udev/9630/23

@zyga The conclusion we came to with respect to how this interface works in conjunction with an interface that requires udev tagging to work properly is that the greengrass-support interface will be the one that "wins" and as such udev + the device cgroup will be disabled even if it's needed for the interface to properly confine because greengrass-support is a super-privileged interface. Additionally, I am going to work with upstream to see how we can eventually support "stacking" the cgroups that greengrass enables inside the cgroup that snapd enables.

zyga commented on interfaces/udev/spec.go:
```diff
@@ -73,10 +81,12 @@ func udevTag(securityTag string) string {
 // TagDevice adds an app/hook specific udev tag to devices described by the
 // snippet and adds an app/hook-specific RUN rule for hotplugging.
 func (spec *Specification) TagDevice(snippet string) {
```
Nitpick, `if spec.controlsDeviceCgroup { return }`

anonymouse64 replied:
gofmt says I should use ```go
func (spec *Specification) TagDevice(snippet string) {
	if spec.controlsDeviceCgroup {
		return
	}
	for _, securityTag := range spec.securityTags {
	...
```

is this okay or do you want it on one line?

anonymouse64 commented:
I looked at snap-confine, and the only place where snap-confine will setup udev (and thus cgroups) is here: https://github.com/snapcore/snapd/blob/master/cmd/snap-confine/snap-confine.c#L354, and as such as long as we don't define udev tagging from snapd there shouldn't be any problem with NVIDIA specific stuff.

I also empirically tested this on my machine with NVIDIA hardware + drivers and didn't observe any additional udev rules being added so I think this is okay.

anonymouse64 commented:
I refactored the test here so that it actually tests that udev goes away and so that the tests were simpler by using some of the helpers that I assume weren't around when the greengrass-support interface was originally written.

pedronis commented on interfaces/udev/spec.go:
```diff
@@ -73,10 +81,12 @@ func udevTag(securityTag string) string {
-		tag := udevTag(securityTag)
-		spec.addEntry(fmt.Sprintf("# %s\n%s, TAG+=\"%s\"", spec.iface, snippet, tag), tag)
-		spec.addEntry(fmt.Sprintf("TAG==\"%s\", RUN+=\"/usr/lib/snapd/snap-device-helper $env{ACTION} %s $devpath $major:$minor\"", tag, tag), tag)
+	if !spec.controlsDeviceCgroup {
```
this check is too early because at this point interfaces can be considered in any order so the flag might not be set yet though it's relevant,  we need to consider the flag  in the backend deriveContent or so and not at the point we are inside repo.SnapSpecification.   The way supressHomeIx works in the apparmor backend can give some inspiration.

anonymouse64 replied:
I can certainly move it to there, but this does complicate testing as `deriveContent` would be the function of the udev backend to test, but it's not currently exported, so building up all the required infrastructure to test that the `Setup` function which calls `deriveContent` seemed excessive and complex to me, when all it's really testing is that the original `spec.Snippets` method returns nothing. This is why I chose to implement this logic in `spec.Snippets` instead, as it's easier to test.

Note that technically the result is the same whether we do the check in `deriveContent` or in `Snippets`.

pedronis replied:
I'm a bit confused, this bit of code is not in Snippets at all, it's too early as I said, we can try to catch up still today about this

anonymouse64 replied:
Ah sorry I read the location of this comment wrong. This check is both in `TagDevice` and in `Snippets`. It's redundant to have it both places, but I thought it would be more efficient to have it here in `TagDevice` in addition to `Snippets` as that minimizes the strings we store using `addEntry`, but it's minimal savings so it's probably not a big deal to have it here as well.

pedronis replied:
As I tried to say the issue with doing this only here is that is potentially not correct, it produces a behiavor that depends on the order interfaces are considered.

anonymouse64 replied:
I will remove it here, but what about the equivalent check in `Snippets`?

anonymouse64 commented:
I rebased this for clarity, and also renamed some of the methods to align with how the apparmor suppress home ix and apparmor uses ptrace trace methods are designed.

pedronis commented on interfaces/udev/spec.go:
```diff
@@ -93,6 +108,14 @@ func (c byTagAndSnippet) Less(i, j int) bool {
 
 // Snippets returns a copy of all the snippets added so far.
 func (spec *Specification) Snippets() (result []string) {
+	// If the interface controls it's own device cgroup, then
```
s/the interface/one of the interfaces/ , spec is used over all the plugs and slots of a snap.

anonymouse64 replied:
Done

pedronis commented on interfaces/udev/spec.go:
```diff
@@ -93,6 +108,14 @@ func (c byTagAndSnippet) Less(i, j int) bool {
+	// there are udev rules, and as such we don't want to generate any udev
+	// rules
+
+	if spec.ControlsDeviceCgroup() {
```
this merits  a backend_test.go  test  with a snap  having plugs of a test interface that would TagDevice and one that declares to control the device cgroup.

anonymouse64 replied:
I added a test here similar to the other tests in backed_test.go that just checks the udev rules file doesn't exist when `stat`'d.

pedronis commented on interfaces/udev/backend_test.go:
```diff
@@ -322,6 +323,22 @@ func (s *backendSuite) TestCombineSnippetsWithActualSnippets(c *C) {
+	// NOTE: Hand out a permanent snippet so that .rules file is generated.
+	s.Iface.UDevPermanentSlotCallback = func(spec *udev.Specification, slot *snap.SlotInfo) error {
+		spec.SetControlsDeviceCgroup()
+		spec.AddSnippet("dummy")
```
the test in this style is probably more interesting if you switch the order of these two, first AddSnippet then Set...

anonymouse64 replied:
Done

pedronis commented on interfaces/udev/backend_test.go:
```diff
@@ -322,6 +323,22 @@ func (s *backendSuite) TestCombineSnippetsWithActualSnippets(c *C) {
+		snapInfo := s.InstallSnap(c, opts, "", ifacetest.SambaYamlV1, 0)
+		fname := filepath.Join(dirs.SnapUdevRulesDir, "70-snap.samba.rules")
+		_, err := os.Stat(fname)
+		c.Assert(err, ErrorMatches, fmt.Sprintf("stat.*%s.*no such file or directory", fname))
```
I think we have a testutil.FileAbsent checker nowadays (is not very used yet) for this, or also osutil.FileExists

anonymouse64 replied:
Great, thanks for the tip! I used `testutil.FileAbsent` here instead as that's a more direct and clear test

jdstrand commented on interfaces/builtin/greengrass_support_test.go:
```diff
@@ -80,19 +87,31 @@ func (s *GreengrassSupportInterfaceSuite) TestSanitizeSlot(c *C) {
 
 func (s *GreengrassSupportInterfaceSuite) TestSanitizePlug(c *C) {
 	c.Assert(interfaces.BeforePreparePlug(s.iface, s.plugInfo), IsNil)
+
```
Please remove this blank line.

anonymouse64 replied:
Done

jdstrand commented on interfaces/builtin/greengrass_support_test.go:
```diff
@@ -80,19 +87,31 @@ func (s *GreengrassSupportInterfaceSuite) TestSanitizeSlot(c *C) {
-	c.Assert(apparmorSpec.SecurityTags(), DeepEquals, []string{"snap.other.app2"})
-	c.Check(apparmorSpec.SnippetForTag("snap.other.app2"), testutil.Contains, "mount fstype=overlay no_source -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/**,\n")
+func (s *GreengrassSupportInterfaceSuite) TestSecCompSpec(c *C) {
+	spec := &apparmor.Specification{}
```
The name of this test should be TestAppArmorSpec().

anonymouse64 replied:
Thanks for catching this, fixed

jdstrand commented on interfaces/builtin/greengrass_support_test.go:
```diff
@@ -80,19 +87,31 @@ func (s *GreengrassSupportInterfaceSuite) TestSanitizeSlot(c *C) {
+	c.Check(spec.SnippetForTag("snap.other.app2"), testutil.Contains, "mount fstype=overlay no_source -> /var/snap/{@{SNAP_NAME},@{SNAP_INSTANCE_NAME}}/**,\n")
+}
+
+func (s *GreengrassSupportInterfaceSuite) TestAppArmorSpec(c *C) {
```
The name of this test should be TestSecCompSpec().

anonymouse64 replied:
Thanks for catching this, fixed

jdstrand commented on interfaces/builtin/greengrass_support_test.go:
```diff
@@ -80,19 +87,31 @@ func (s *GreengrassSupportInterfaceSuite) TestSanitizeSlot(c *C) {
+
+	// connect the greengrass-support interface and ensure the spec is now nil
+	c.Assert(spec.AddConnectedPlug(s.iface, s.plug, s.slot), IsNil)
+	c.Check(spec.Snippets(), HasLen, 0)
```
It might be interesting to also have a test where greengrass-support is connected and then you connect network-control and show that the spec is still nil.

anonymouse64 replied:
Done

jdstrand commented on interfaces/udev/backend_test.go:
```diff
@@ -322,6 +322,21 @@ func (s *backendSuite) TestCombineSnippetsWithActualSnippets(c *C) {
+	for _, opts := range testedConfinementOpts {
+		snapInfo := s.InstallSnap(c, opts, "", ifacetest.SambaYamlV1, 0)
+		fname := filepath.Join(dirs.SnapUdevRulesDir, "70-snap.samba.rules")
+		c.Check(fname, testutil.FileAbsent)
```
Nice test.

jdstrand commented on interfaces/udev/spec.go:
```diff
@@ -93,6 +108,14 @@ func (c byTagAndSnippet) Less(i, j int) bool {
+	}
 	entries := make([]entry, len(spec.entries))
 	copy(entries, spec.entries)
 	sort.Sort(byTagAndSnippet(entries))
```
It would be good to add some tests to udev/spec_test.go (eg, like we did for suppressHomeIx).

anonymouse64 replied:
Done

jdstrand commented on interfaces/builtin/common.go:
```diff
@@ -162,8 +163,14 @@ func (iface *commonInterface) SecCompConnectedPlug(spec *seccomp.Specification,
+			spec.TagDevice(rule)
+		}
 	}
+
```
It would be good to add tests to common_test.go like we did for suppressHomeIx.

anonymouse64 replied:
Done
